app/controllers/api/v1/projects_controller.rb

```diff
@@ -26,32 +26,54 @@ def show
       def create
         raise PermissionError.new(Project, :create) unless can_create_projects?(@team)
 
-        project = @team.projects.build(project_params.merge!(created_by: current_user))
+        ActiveRecord::Base.transaction do
+          project = @team.projects.build(project_params.merge!(created_by: current_user))
+
+          project.save!
+
+          if project_params[:visibility] == 'visible'
+            project.team_assignments.create!(
+              team: project.team,
+              user_role: UserRole.find_predefined_viewer_role,
+              assigned_by: current_user,
+              assigned: :manually
+            )
+          end
 
-        if project.visible? # set default viewer role for public projects
-          project.default_public_user_role = UserRole.predefined.find_by(name: I18n.t('user_roles.predefined.viewer'))
+          render jsonapi: project, serializer: ProjectSerializer, scope: { metadata: params['with-metadata'] == 'true' }, status: :created
         end
-
-        project.save!
-
-        render jsonapi: project, serializer: ProjectSerializer, scope: { metadata: params['with-metadata'] == 'true' }, status: :created
       end
 
       def update
         @project.assign_attributes(project_params)
+        has_team_assignment = @project.team_assignments.where(team: @team).any?
```

Comment:
it can be loaded later, after the first guard clause, and it will be more efficient to load an object itself to avoid multiple database queries